### ISSUE TITLE
Why clamp inputs to all tanh calls?

### ISSUE THREAD

**Issue by dhruvdcoder** [dalab/hyperbolic_nn#1]
https://github.com/dalab/hyperbolic_nn/blob/45be2f6698706bba48bb015e7a8ed32408d0cf6e/util.py#L26

We are trying to reimplement the layers proposed by the [Hyperbolic Neural Networks]( https://arxiv.org/abs/1805.09112) paper. We use float64 instead of float32 for the entire model and inputs. Hence, we avoid numerical instability. However, if we do not clamp the inputs to the tanh functions between (-15, 15), the network does not seem to train at all. It would be great if you could provide a reason for doing this and for picking the value of 15. 

PS: I really liked the paper and thank you for making the code available.

**Comment by octavian-ganea** [comment]
I am not fully sure, but I believe the reason is that exp(15) is very large and exp(-15) is very small, so, in this range, small updates in the exponent can result in very large/tiny fluctuations which can lead to numerical instabilities, overflow/underflow. Moreover, gradient flow might be very noisy (for exp(15)) or 0 (for exp(-15)), which can mean learning signal is very poor. Anyways, MAXINT32 ~= exp(21).

Using float64 would help, but would double the training speed ....

One solution to this problem is to always gyro-translate your embeddings to have 0 mean. Or increase dimensionality (see https://arxiv.org/abs/1804.03329 ).
